File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        min=float('inf')
        visited=[[0]*len(grid[0]) for _ in range(len(grid))]
        rows=len(grid)
        cols=len(grid[0])
        q=deque()
        fresh=0
        for row in range(rows):
            for col in range(cols):
                if not visited[row][col] and grid[row][col]==2:
                    visited[row][col]=1
                    q.append((row,col,0))
                if grid[row][col]==1:
                    fresh+=1
        adjacent=[(0,1),(1,0),(0,-1),(-1,0)]
        lasttime=0
        while (len(q)>0):
            currow,curcol,time=q.popleft()
            for drow , dcol in adjacent:
                adjrow,adjcol=currow+drow,curcol+dcol
                if 0<=adjrow<rows and 0<=adjcol<cols and not visited[adjrow][adjcol] and grid[adjrow][adjcol]==1:
                    visited[adjrow][adjcol]=1
                    grid[adjrow][adjcol]=2
                    fresh-=1
                    q.append((adjrow,adjcol,time+1))
                    lasttime=time+1
        if fresh==0:
            return lasttime
        return -1

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int rows=grid.size();
        int cols=grid[0].size();
        queue<vector<int>> q;
        int count=0;
        for (int i=0;i<rows;i++){
            for (int j=0;j<cols;j++){
                if (grid[i][j]==2){
                    q.push({i,j});
                }
            }
        }
        vector<vector<int>> adj={{0,1},{0,-1},{1,0},{-1,0}};
        while (!q.empty()){
            int k=q.size();
            for (int i=0;i<k;i++){
                vector<int> current=q.front();
                q.pop();
                for (vector<int> a : adj){
                    int x=current[0]+a[0];
                    int y=current[1]+a[1];
                    if (x>=0 && x <rows && y>=0 && y<cols && grid[x][y]==1){
                        grid[x][y]=2;
                        q.push({x,y});
                    }
                }
            }
            count++;
        }
        for (int i=0;i<rows;i++){
            for (int j=0;j<cols;j++){
                if (grid[i][j]==1){
                    return -1;
                }
            }
        }
        if (count==0){
            return count;
        }
        return count-1;
        

    }
};